src/snowflake/snowpark/ml/transformer.py

```diff
@@ -0,0 +1,360 @@
+            if self._input_cols
+            else None
+        )
+        self._states_table_name = random_name_for_temp_object(TempObjectType.TABLE)
+        self._category_table_name = {}
+        self._encoder_count_table = random_name_for_temp_object(TempObjectType.TABLE)
+
+    @property
+    def input_cols(self) -> List[str]:
+        return self._input_cols
+
+    @input_cols.setter
+    def input_cols(self, value: List[str]):
+        self._input_cols = value
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+        self._encoder_table_cols = (
+            [f"states_{c}_encoder" for c in self._input_cols]
+            if self._input_cols
+            else None
+        )
+
+    @property
+    def output_cols(self) -> List[str]:
+        return self._output_cols
+
+    @output_cols.setter
+    def output_cols(self, value: List[str]):
+        self._output_cols = value
+
+    def fit(self, df: DataFrame) -> "OneHotEncoder":
+        if self._output_cols is None or len(self._output_cols) == 1:
+            self.merge = True
+        elif len(self._input_cols) == len(self._output_cols):
+            self.merge = False
+        else:
+            raise ValueError("number of output colum and input column does not match")
+        encoder_fitter(self, df, "onehot")
+        return self
+
+    def transform(
+        self,
+        df: DataFrame,
+        sparse: bool = False,
+    ) -> DataFrame:
+        originalC = df.columns
+        encoder_structure = array_construct()
+        encoder_offset = lit(0)
+        df_count_table = df._session.table(self._encoder_count_table)
+        df = df.join(df_count_table)
+        for fc, ec, (c, tabname) in zip(
+            self._states_table_cols,
+            self._encoder_table_cols,
+            self._category_table_name.items(),
+        ):
+            cattable = df._session.table(tabname)
+            if sparse:
+                sparse_array = cattable.with_column("sparse", lit(0)).select(
+                    [array_agg(col("sparse")).as_(f"{fc}_sparse")]
+                )
+                sparse_array = cattable.join(sparse_array).join(df_count_table)
+                cattable = sparse_array.select(
+                    [
+                        col(fc),
+                        array_slice(
+                            array_insert(col(f"{fc}_sparse"), col(ec), lit(1)),
+                            lit(0),
+                            col(f"{fc}_count"),
+                        ).as_(ec),
+                    ]
+                )
+            df = cattable.join(df, df[c] == cattable[fc])
+            if not sparse:
+                encoder_structure = array_append(
+                    encoder_structure, col(ec) + encoder_offset
+                )
+                encoder_offset += col(f"{fc}_count")
+            else:
+                encoder_structure = array_cat(encoder_structure, col(ec))
```

Comment:
Why do you use `array_cat` here?

Reply:
if customer want the sparse form, col(ec) here would be a one hot array of the corresponding input col and encoder_structure is only used when you want to merge it.